tao_compiler/mlir/disc/transforms/disc_dense_to_sparse.cc

```diff
@@ -0,0 +1,428 @@
+      }
+      auto newInputTy =
+          RankedTensorType::get(inputTranShape, inputTy.getElementType());
+
+      auto inputTranOp = InsertTranspose(op, input, transposeAttr, b);
+      inputTranOp->getResult(0).setType(newInputTy);
+      newOperands.push_back(inputTranOp->getResult(0));
+    } else {
+      newOperands.push_back(input);
+    }
+
+    newOperands.push_back(kernel);
+
+    // compute spgemm output shape
+    auto output_shape = outputTy.getShape();
+    llvm::SmallVector<int64_t, 4> spgemm_shape;
+    if (sparse_weight_pos == 1) {
+      for (int64_t val : transposeAttr) {
+        spgemm_shape.push_back(output_shape[val]);
+      }
+    } else {
+      for (int64_t val : output_shape) {
+        spgemm_shape.push_back(output_shape[val]);
+      }
+    }
+
+    auto newOutputTy =
+        RankedTensorType::get(spgemm_shape, outputTy.getElementType());
+
+    // insert spgemm op
+    auto dot_dimension_numbers = op.dot_dimension_numbers();
+    auto lhs_contracting_dimensions =
+        dot_dimension_numbers.getLhsContractingDimensions();
+    auto rhs_contracting_dimensions =
+        dot_dimension_numbers.getRhsContractingDimensions();
+
+    int64_t lhs_cd, rhs_cd;
+    if (sparse_weight_pos == 1) {
+      lhs_cd = 1 - lhs_contracting_dimensions[0];
+      rhs_cd = rhs_contracting_dimensions[0];
+    } else {
+      lhs_cd = rhs_contracting_dimensions[0];
+      rhs_cd = lhs_contracting_dimensions[0];
+    }
+
+    auto lhs_cd_attr = b.getNamedAttr("lhs_contracting_dimensions",
+                                      b.getI64IntegerAttr(lhs_cd));
+    auto rhs_cd_attr = b.getNamedAttr("rhs_contracting_dimensions",
+                                      b.getI64IntegerAttr(rhs_cd));
+
+    NamedAttribute attrs[] = {lhs_cd_attr, rhs_cd_attr};
+    auto config = b.getDictionaryAttr(attrs);
+
+    auto spgemm = b.create<mhlo_disc::CustomCallOp>(
+        loc, newOutputTy, newOperands, "sparse_gemm", false, config);
+
+    spgemm.getResult(0).setType(newOutputTy);
+
+    // insert transpose after spgemm when sparse_weight_pos = 1
+    if (sparse_weight_pos == 1) {
+      auto output_transpose_op =
+          InsertTranspose(op, spgemm.getResult(0), transposeAttr, b);
+      output_transpose_op->getResult(0).setType(outputTy);
+      rewriter.replaceOp(op, output_transpose_op->getResult(0));
+    } else {
+      rewriter.replaceOp(op, spgemm.getResult(0));
+    }
+    return success();
+  }
+
+  LogicalResult extractWeightFromConst(mhlo::ConstantOp constOp,
+                                       bool is_transpose) const {
+    if (!constOp) return failure();
+
+    auto constTy = constOp.getResult().getType().dyn_cast<RankedTensorType>();
+    if (!constTy || !constTy.hasStaticShape() ||
+        !constTy.getElementType().isF16())
+      return failure();
+
+    auto weight_type = constOp.getType().cast<ShapedType>();
+    ArrayRef<int64_t> weight_shape = weight_type.getShape();
```

Comment:
please check the rank of the weight is two since dot_general supports other cases. And maybe also need to check the non-const operand as well if you only want to support 2dx2d case.

Reply:
Done